drivers/timer/cdns/timer.c

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright 2024, UNSW
+ * Copyright 2025, UNSW
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <stdint.h>
+#include <microkit.h>
+#include <sddf/timer/protocol.h>
+#include <sddf/util/util.h>
+#include <sddf/util/printf.h>
+#include <sddf/resources/device.h>
+
+__attribute__((__section__(".device_resources"))) device_resources_t device_resources;
+
+/*
+ * Zynq UltraScale+ MPSoC contains a timer (Cadence Triple Timer Clock) with 3 32-bit counters.
+ * Only 2 of the 3 timers are used in this driver. All 3 timers by default are connected to
+ * the LSBUS clock in the Low Power Domain, (controlled by LPD_LSBUS_CTRL).
+ */
+/* TODO:
+ * More graceful error when user requests timeout greater than served by ZCU102_TIMER_MAX_TICKS (at 25MHz (downscaled), 32-bit counter thats ~2min 51.799s)
```

Comment:
Shouldn't the driver just handle this kind of timeout by waiting for multiple hardware-level timeouts? I thought that's how the other drivers worked.

Reply:
I was looking for how to handle this, and went for the most simple solution based on how jh7110 driver handles this case: https://github.com/au-ts/sddf/blob/main/drivers/timer/jh7110/timer.c#L132

I can rewrite it to handle longer timeouts as you've suggested for both the drivers, just some more logic will be needed.

Reply:
Okay I guess I was lazy with that timer driver.

Reply:
I'll add a fix for that in both then, shouldn't be too difficult.

Reply:
Fixed in: https://github.com/au-ts/sddf/pull/407/commits/7ba92672f953f3d2eedd65b60b4d38cb2eaaae13